File: algorithms/algorithms-cplusplus/leetcode/Question_0647_Palindromic_Substrings.cc
#include <string>

using namespace std;

class Solution {
 public:
  int countSubstrings(string s) {
    int res = 0;
    for (int i = 0; i < s.size(); i++) {
      res += help(s, i, i);
      res += help(s, i, i + 1);
    }
    return res;
  }

  int help(string s, int l, int r) {
    int res = 0;
    while (l >= 0 && r < s.size() && s[l] == s[r]) {
      res++;
      l--;
      r++;
    }
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0647_Palindromic_Substrings.java
package com.lanceolata.coding.algorithms.java.leetcode;

public class Question_0647_Palindromic_Substrings {
    public int countSubstrings(String s) {
        if (s == null || s.length() == 0) {
            return 0;
        }
        int res = 0;
        for (int i = 0; i < s.length(); i++) {
            res += extendPalindrome(s, i, i);
            res += extendPalindrome(s, i, i + 1);
        }
        return res;
    }

    private int extendPalindrome(String s, int l, int r) {
        int res = 0;
        while (l >= 0 && r < s.length() && s.charAt(l) == s.charAt(r)) {
            res++;
            l--;
            r++;
        }
        return res;
    }
}
